components/tooltip/script.js

```diff
@@ -90,15 +107,49 @@ class Tooltip extends HTMLElement {
                 break;
             default:
                 position.top = elementSize.top - TOOLTIP_MARGIN - tooltipSize.height;
-                console.log(`The provided option for position ${this.position} is not valid - using top as a fallback. Possible options are top, bottom, left and right.`);
-                this.position = 'top';
+                console.log(`The provided option for position ${orientation} is not valid - using top as a fallback. Possible options are top, bottom, left and right.`);
+                orientation = 'top';
                 break;
         }
 
-        this.style.top = position.top + 'px';
-        this.style.left = position.left + 'px';
-        this.style.visibility = 'visible';
-        this.visible = true;
+        this.position = orientation;
+
+        this.style.top = scrollOffsetY + position.top + 'px';
+        this.style.left = scrollOffsetX + position.left + 'px';
+
+        // check if the tooltip still overflows and re-position it again
+        const overflows  = this.overflows();
+        if(Object.keys(overflows).length && this.fixedSides.length !== 4) {
+            this.setPosition(this.getVisiblePosition(overflows));
+        }
+    }
+
+    overflows() {
+        var rect = this.getBoundingClientRect();
+        const overflows = {};
+
+        if (rect.top < 0) overflows.top = true;
+        if (rect.left < 0)  overflows.left = true;
+        if (rect.right > (window.innerWidth || document.documentElement.clientWidth)) overflows.right = true;
+        if (rect.bottom > (window.innerHeight || document.documentElement.clientHeight)) overflows.bottom = true;
+
+        return overflows;
+    }
+
+    getVisiblePosition(overflows) {
+        const allSides = ['top', 'bottom', 'left', 'right'];
+
+        overflows = overflows || this.overflows();
+        let overflowingSides = Object.keys(overflows);
+
+        overflowingSides = overflowingSides.filter(side => this.fixedSides.indexOf(side) == -1);
+
+        let nextIndex = allSides.indexOf(this.position) + 1;

```

Comment:
Maybe you wanted to do `overflowingSides` instead of `allSides`?

Reply:
Actually, here we don't need overflowingSides at all. We check if the tooltip overflows and then we try one of the other possible positions. I updated the method a little to only try the positions that are not currently overflowing or that haven't been tried yet. This reduced the amount of tries.